include/zephyr/drivers/emul.h

```diff
@@ -83,10 +81,10 @@ extern const struct emul __emul_list_end[];
  * @param data_ptr emulator-specific data
  */
 #define EMUL_DEFINE(init_ptr, node_id, cfg_ptr, data_ptr)                                          \
-	static struct emul EMUL_REG_NAME(node_id) __attribute__((__section__(".emulators")))       \
+	const struct emul EMUL_REG_NAME(node_id) __attribute__((__section__(".emulators")))        \
 	__used = {                                                                                 \
 		.init = (init_ptr),                                                                \
-		.dev_label = DT_LABEL(node_id),                                                    \
+		.name = DT_LABEL(node_id),                                                         \
```

Comment:
nit: extra whitespace at end here.

Do we need `name` at all?  If so, can we use `Z_DEVICE_DT_DEV_NAME`

Reply:
emulators are defined independently of devices (not sure the reasons), so unless we collect it here there's no way to retrieve it (we can't do like `dev->name`).

Reply:
The reason for emulators to be defined independently is that they use the same node in the devicetree. So they would have the same label which would cause a conflict with calling `device_get_binding`. The dts will look like:

```
/ {
  &i2c0 {
    my_dev: my_dev@80 {
      ...
    };
  };
};
```

Doing this means we can use the exact same dts file for both the real board, and only need to change the compatible string of the `i2c0` bus for the host test. The device will then generate both a normal device driver AND an emulator using the same label and compatible string.

Reply:
> The reason for emulators to be defined independently is that they use the same node in the devicetree. So they would have the same label which would cause a conflict with calling `device_get_binding`. The dts will look like:
> 
> ```
> / {
>   &i2c0 {
>     my_dev: my_dev@80 {
>       ...
>     };
>   };
> };
> ```
> 
> Doing this means we can use the exact same dts file for both the real board, and only need to change the compatible string of the `i2c0` bus for the host test. The device will then generate both a normal device driver AND an emulator using the same label and compatible string.

I still don't understand why device infrastructure can't be used (or at least re-used). Could you provide a real example (and how it is used) I can look at?

Reply:
Agreed, I'm confused by the statement of `emulators are defined independently`.  What exactly do you mean.

How or when don't we have unique devicetree nodes for the emulated devices?

Reply:
Why do `struct emul` need a `dev_label`, why can't it use `struct device *`

Reply:
> Agreed, I'm confused by the statement of emulators are defined independently. What exactly do you mean.

> Why do `struct emul` need a `dev_label`, why can't it use `struct device *`

I started typing up a reply that got longer than I wanted. I'll try to be concise here, but I think I can describe this unambiguously at length. I'll avoid unqualified use of the word "device," because it seems to be overloaded in this context. I will use "component" to refer to a physical part on a board, because I don't think that word is being used for anything else here. "Emulator" also seems to be overloaded, and I don't have a great solution for that. I'll use it here to refer to a `struct emul` defined in `zephyr/subsys/emul` or the code that contains such a definition.

A `struct device` holds the runtime state for a device driver instance. In a typical configuration with devicetree, there would be one `struct device`/driver instance for each node in the tree. This is true for a production build on real hardware and for a test build with simulated hardware.

A `struct emul` holds the runtime state for an emulator instance, where an emulator instance corresponds to a physical component that would exist on a production board. In a typical test configuration using emulators, there would be `struct emul`s corresponding to several nodes in the tree. For the nodes in use, there is a `struct emul` (representing a component) and also a `struct device` (driver instance).

> I still don't understand why device infrastructure can't be used (or at least re-used). Could you provide a real example (and how it is used) I can look at?

Consider `bmi160.c`, `emul_bmi160.c`, and `tests/drivers/sensor/accel/boards/native_posix.overlay`. These are a driver for the BMI160 component, an emulator for that component, and a test that uses that emulator to exercise that driver. All 3 of these use devicetree nodes with the compatible string "bosch,bmi160". There are a `struct emul` and a `struct device` corresponding to the same DT node. AFAICT, this is the pattern in all of the existing tests in this repo that use emulators.

> I still don't understand why emulators are not regular devices. It looks like `gpio_emul.c` is, or `eeprom_emulator.c`.

These files define device drivers, but the drivers and their DT nodes don't correspond to any physical component. I'm not familiar with this code, but it looks like it's meant to be used when tests conceptually need GPIO or EEPROM, but the test isn't supposed to exercise any particular GPIO controller driver or EEPROM driver. I see a few similar examples in `bbram_emul.c` and `adc_emul.c`, which are both drivers.

These types of drivers are useful, but you can't use them to exercise the driver for a specific BBRAM or ADC component. To do that, you would need an emulator, as in a `struct emul`. You _could_ just directly simulate the I2C traffic in a driver unit test, but that gets pretty tedious in a large test for a complicated driver. `struct emul` provides a useful encapsulation for the behavior of a component.

*tl;dr:* `struct device` and `struct emul` represent distinct concepts. It's appropriate for a `struct device` and a `struct emul` to be based on the same DT node.